### PR TITLE
KeycloakAuthenticator: configurable roles retrieval

### LINKED ISSUE #35: Problems with get_roles_for_token
We trying to use this module for our keycloak/jupyter authentication. This module works well in a setup where roles are defined in the keycloak. However, we dynamically update user attributes with role names (List<String>). Then we use Client Mapper to pick multivalued User Attribute. Here we pass the roles to KeycloakAuthenticator.

As soon as we create new "Mapper" in Client, it overwrites all the resource_access and it removes existing system roles too. It takes out parameters which are used by other parts of jupyter such as resource_access.broker.token(its used for provider token access). I can see why it is doing it, because the newly created mapper has higher priority and it uses Token Claim Name resouce_access.{client_id}.roles. One solution could be to read the roles from an alternative Token Claim name such as roles_claim_key. 

Is it possible to provide an update to have a roles_claim_key as configurable parameter; the default can be resource_access.{client_id}.roles . However, if we should be allowed change to  for example custom.roles. This would be massively help us.

This functionality exists on 

https://github.com/jupyterhub/oauthenticator/blob/f830d4b790cd2fa6e5a2f5d6ae9b1ee844490944/oauthenticator/generic.py#L23

but its for groups not for roles.

I think the below function needs updating; where you can pass either token or custom like : roles_claim_key

https://github.com/swan-cern/jupyterhub-extensions/blob/c250978c7f2666c0caca7861d98c0747574f218f/KeyCloakAuthenticator/keycloakauthenticator/auth.py#L149

    def get_roles_for_token(self, token):
        decoded_token = self._decode_token(token)
        return set(
            decoded_token.\
               get('resource_access', {'app': ''}).\
               get(self.client_id, {'roles_list': ''}).\
               get('roles', 'no_roles')
        )

Thanks for your help.

### PR DESCRIPTION
**Pull request by diocas**
Replace accepted_roles by allowed_roles and make it a list
Validate callables

Closes #35 
Replaces #36

### CODE REVIEW

**Comment by diocas** [comment]
@rmalik2018 as I mentioned in a previous message, I decided to do something a bit more generic that what you proposed. Can you check this is ok for you?

File: KeyCloakAuthenticator/keycloakauthenticator/auth.py [hunk 1]
```diff
@@ -93,11 +105,35 @@ def pre_spawn_hook(authenticator, spawner, auth_state):
...
+                get(env.client_id, {'roles_list': ''}).\
+                get('roles', []))
+        return get_roles
+
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
 
         # Force auth state so that we can store the tokens in the user dict
         self.enable_auth_state = True
+        self._allowed_roles = set(self.allowed_roles)
```
Comment (etejedor) [inline KeyCloakAuthenticator/keycloakauthenticator/auth.py:136]: So we change the type of `allowed_roles` from set to list, and then we create a mirror attribute which is a set?

Comment (diocas) [reply KeyCloakAuthenticator/keycloakauthenticator/auth.py:136]: Yes. I do this to allow the intersection of 2 sets (when checking if a user has the necessary roles to be allowed in). Do you have a better suggestion?

Comment (etejedor) [reply KeyCloakAuthenticator/keycloakauthenticator/auth.py:136]: Then why changing the type of the `allowed_roles` attribute in the first place?

Comment (etejedor) [reply KeyCloakAuthenticator/keycloakauthenticator/auth.py:136]: Also, once `allowed_roles` is set it won't change, will it? Meaning `allowed_roles` and `_allowed_roles` do not need to be synchronized.

**Review by etejedor — APPROVED** [review]

File: KeyCloakAuthenticator/keycloakauthenticator/auth.py (hunk 1 above)
Comment (diocas) [reply KeyCloakAuthenticator/keycloakauthenticator/auth.py:136]: I'm assuming no, hope I'm not wrong...

I'm changing the attribute to ensure we can configure it via yaml files in the k8s deployment (and that doesn't support sets).

**Review by rccern — DISMISSED** [review]

**Comment by diocas** [comment]
Just added a comment to the code and fixed the description of one variable. Merging.

**Pull request merged**